src/hotspot/share/gc/shenandoah/shenandoahGenerationalFullGC.cpp

```diff
@@ -0,0 +1,388 @@
+  if (_old_to_region != nullptr) {
+    log_debug(gc)("Planned compaction into Old Region " SIZE_FORMAT ", used: " SIZE_FORMAT " tabulated by worker %u",
+            _old_to_region->index(), _old_compact_point - _old_to_region->bottom(), _worker_id);
+    _old_to_region->set_new_top(_old_compact_point);
+    _old_to_region = nullptr;
+  }
+}
+
+void ShenandoahPrepareForGenerationalCompactionObjectClosure::finish_young_region() {
+  if (_young_to_region != nullptr) {
+    log_debug(gc)("Worker %u planned compaction into Young Region " SIZE_FORMAT ", used: " SIZE_FORMAT,
+            _worker_id, _young_to_region->index(), _young_compact_point - _young_to_region->bottom());
+    _young_to_region->set_new_top(_young_compact_point);
+    _young_to_region = nullptr;
+  }
+}
+
+bool ShenandoahPrepareForGenerationalCompactionObjectClosure::is_compact_same_region() {
+  return (_from_region == _old_to_region) || (_from_region == _young_to_region);
+}
+
+void ShenandoahPrepareForGenerationalCompactionObjectClosure::do_object(oop p) {
+  assert(_from_region != nullptr, "must set before work");
+  assert((_from_region->bottom() <= cast_from_oop<HeapWord*>(p)) && (cast_from_oop<HeapWord*>(p) < _from_region->top()),
+         "Object must reside in _from_region");
+  assert(_heap->complete_marking_context()->is_marked(p), "must be marked");
+  assert(!_heap->complete_marking_context()->allocated_after_mark_start(p), "must be truly marked");
+
+  size_t obj_size = p->size();
+  uint from_region_age = _from_region->age();
+  uint object_age = p->age();
+
+  bool promote_object = false;
+  if ((_from_affiliation == ShenandoahAffiliation::YOUNG_GENERATION) &&
+      (from_region_age + object_age >= _tenuring_threshold)) {
+    if ((_old_to_region != nullptr) && (_old_compact_point + obj_size > _old_to_region->end())) {
+      finish_old_region();
+      _old_to_region = nullptr;
+    }
+    if (_old_to_region == nullptr) {
+      if (_empty_regions_pos < _empty_regions.length()) {
+        ShenandoahHeapRegion* new_to_region = _empty_regions.at(_empty_regions_pos);
+        _empty_regions_pos++;
+        new_to_region->set_affiliation(OLD_GENERATION);
+        _old_to_region = new_to_region;
+        _old_compact_point = _old_to_region->bottom();
+        promote_object = true;
+      }
+      // Else this worker thread does not yet have any empty regions into which this aged object can be promoted so
+      // we leave promote_object as false, deferring the promotion.
+    } else {
+      promote_object = true;
+    }
+  }
+
+  if (promote_object || (_from_affiliation == ShenandoahAffiliation::OLD_GENERATION)) {
+    assert(_old_to_region != nullptr, "_old_to_region should not be nullptr when evacuating to OLD region");
+    if (_old_compact_point + obj_size > _old_to_region->end()) {
+      ShenandoahHeapRegion* new_to_region;
+
+      log_debug(gc)("Worker %u finishing old region " SIZE_FORMAT ", compact_point: " PTR_FORMAT ", obj_size: " SIZE_FORMAT
```

Comment:
This evaluates whether or not logging is set to 'debug' for every single live object, right? Might want to change this into log_develop_debug() so that release builds are not impacted by that (and would not log that). There are more instances of this problem in this method.

Reply:
@earthling-amzn will do this.

Reply:
On a closer look, this only happens when a region is finished so it is not on the order of the number of oops. However, after reviewing other uses of `log_debug` I did find a few that would be better as `log_develop_debug`. Also, there are quite some `log_trace` statements in shReferenceProcessor that probably should be `log_develop_trace`, but I'd do that in a separate, unrelated PR.

https://bugs.openjdk.org/browse/JDK-8338779